include/crypto/event.h

```diff
@@ -0,0 +1,378 @@
+ * dynamically.
+ */
+struct ossl_event_queue_method_st {
+    /*
+     * Initialize the auxilliary data
+     *
+     * Expected to return 1 on success, -1 on failure.
+     */
+    int (*init_data)(struct ossl_event_queue_st *queue);
+
+    /*
+     * Destroy the auxilliary data
+     *
+     * Expected to return 1 on success, -1 on failure.
+     */
+    int (*destroy_data)(struct ossl_event_queue_st *queue);
+
+    /*
+     * For a multi-priority queue implementation, this function helps
+     * determine in which event stack the event should end up in.
+     *
+     * Expected to return the priority (which may be zero), or -1 on error
+     */
+    int (*determine_priority)(struct ossl_event_st *event);
+
+    /*
+     * Whenever events have been dispatched to the subscribers, the diverse
+     * return values are counted, and this function is used to determined
+     * how to act on that result.  The returned value is simply returned
+     * back to whoever is calling.
+     *
+     * |count_errors| is the count of any subscriber call that returned
+     * |ossl_event_error|.
+     * |count_success| is the count of any subscriber call that returned
+     * |ossl_event_pass| or |ossl_event_handled|.
+     *
+     * Expected to return 1 for good (may continue to be used), -1 for bad
+     * (this queue is not functional any more).
+     */
+    int (*determine_state)(size_t count_errors, size_t count_success);
+
+    /*
+     * Set / get values to / from auxilliary queue method data, or perform
+     * other queue method specific tasks.
+     *
+     * The functionality this implements is open ended, and allows
+     * extraordinary operations such as reordering (rescheduling) events,
+     * affecting other associated queues, as well as modifications or access
+     * of the auxialliary method data, etc etc etc.
+     * Use with care!
+     *
+     * The arguments are of the same style as ioctl.h functions.  The caller
+     * must know exactly what arguments are expected for each |request|.
+     *
+     * Expected to return 1 on success, -1 on failure.
+     */
+    int (*ctrl)(struct ossl_event_queue_st *queue, uint32_t request,
+                ... /* args */);
+    /* va_list variant of the control function */
+    int (*vctrl)(struct ossl_event_queue_st *queue, uint32_t request,
+                 va_list args);
+
+/*
+ * Request the nummber of priorities for this method.
+ * Takes on extra argument:
+ * - a pointer to size_t, where the number of priorities will be stored
+ *
+ * This is a special request; it expects NULL for |queue|, and gives back a
+ * constant number.
+ */
+# define OSSL_EVENT_C_PRIORITIES    0x00000001
+};
+
+/*-
+ * Functions to create and populate events
+ * ---------------------------------------
+ */
+
+/* Allocate / deallocate an event structure dynamically and populate it */
+struct ossl_event_st *ossl_event_new(uint32_t type, void *ctx,
+                                     const void *identifiers,
```

Comment:
I'm not understanding the purpose of the identifies.
Why wouldn't these be included in the event payload?  I.e. they don't see to fit here.

Can you give an example of how they were intended to be used?

Reply:
I can see an event handler wanting to filter the events it handles based on, say, a stream ID (found in `identities`), and then wanting to pass the actual data (the `payload`) to other functionality that doesn't know or care about the identities part.

Furthermore, I imagine that things like connection ID, stream ID and what not sit in a different structure, or at least separate from the payload being passed.  Making them all part of the payload sounds like doing a bunch of copying into new buffers, when we could as well pass pointers into the received packet (as an example).

Reply:
If so, it needs something finer in resolution that a random pointer.

Reply:
??? random?  I was thinking that this is something known by the event publisher as well as the event subscribers, but that the event mechanism itself doesn't need to know

Reply:
I've added a comment regarding the separation in the event structure

Reply:
Okay, can live with this.  Having the filter in the generic code would be a benefit -- as it stands the event handle will have to be called to determine if the event is suitable.

Things like connection ID and stream ID will be used widely enough that we could justify them in the events which let's us filter generically.  I can't think of anything that wouldn't have a connection ID e.g.

Reply:
One connection can have multiple connection IDs. Which might indicate that it might be useful to filter by connection pointer instead of connection ID.